### PR TITLE
fix(functions): fixes an issue where the functions could not be served

### LINKED ISSUE #3854: Presence of a [dev] block in Netlify TOML causes local dev to fail for functions
**Describe the bug**

The presence of a `[dev]` block in `netlify.toml` causes functions to fail with what looks like a zisi error:
```
Request from ::1: GET /.netlify/functions/hello
TypeError: Cannot destructure property 'inputs' of '(intermediate value)' as it is undefined.
    at buildFunction (/Users/jlengstorf/.nvm/versions/node/v16.13.0/lib/node_modules/netlify-cli/src/lib/functions/runtimes/js/builders/zisi.js:39:11)
    at NetlifyFunction.build (/Users/jlengstorf/.nvm/versions/node/v16.13.0/lib/node_modules/netlify-cli/src/lib/functions/netlify-function.js:53:42)
    at FunctionsRegistry.buildFunctionAndWatchFiles (/Users/jlengstorf/.nvm/versions/node/v16.13.0/lib/node_modules/netlify-cli/src/lib/functions/registry.js:67:38)
```

**To Reproduce**

1. Create an empty folder
2. Add a "hello world" TypeScript function
    ```ts
    import { Handler } from "@netlify/functions";
    
    const handler: Handler = async (event, context) => {
      return {
        statusCode: 200,
        body: JSON.stringify({ message: "Hello World" }),
      };
    };
    
    export { handler };
    ```
3. Add a netlify.toml with a `[dev]` block
    ```toml
    [build]
      functions = "netlify/functions"
      publish = "public"
    
    [dev]
      functions = "netlify/functions"
      publish = "public"
    ```
4. Run `ntl dev`
5. Try to execute the function

Removing the `[dev]` block solves the issue

**Configuration**

```toml
[build]
  functions = "netlify/functions"
  publish = "public"

[dev]
  functions = "netlify/functions"
  publish = "public"
```

```bash
  System:
    OS: macOS 10.15.7
    CPU: (8) x64 Intel(R) Core(TM) i7-7820HQ CPU @ 2.90GHz
    Memory: 92.36 MB / 16.00 GB
    Shell: 5.7.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.5 - /usr/local/bin/yarn
    npm: 8.3.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  npmGlobalPackages:
    netlify-cli: 8.1.8
```

**Expected behavior**

**CLI Output**

If applicable, add the CLI output to help explain your problem.

**Additional context**

### PR DESCRIPTION
**Pull request by lukasholzer**
🎉 Thanks for submitting a pull request! 🎉

#### Summary

Fixes #3854

For us to review and ship your PR efficiently, please perform the following steps:

- [x] Open a [bug/issue](https://github.com/netlify/cli/issues/new/choose) before writing your code 🧑‍💻. This ensures we can discuss the changes and get feedback from everyone that should be involved. If you\`re fixing a typo or something that\`s on fire 🔥 (e.g. incident related), you can skip this step.
- [x] Read the [contribution guidelines](../CONTRIBUTING.md) 📖. This ensures your code follows our style guide and
      passes our tests.
- [x] Update or add tests (if any source code was changed or added) 🧪
- [x] Update or add documentation (if features were changed or added) 📝
- [x] Make sure the status checks below are successful ✅

**A picture of a cute animal (not mandatory, but encouraged)**
image

### CODE REVIEW

**Review by eduardoboucas — COMMENTED** [review]
If I understand correctly, this is doing the same check twice: once in the registry when creating the `NetlifyFunction` instance, and again in the constructor of `NetlifyFunction`. Is that necessary?

**Comment by lukasholzer** [comment]
@eduardoboucas yes you are absolutely right this wasn't the most elegant approach 😂  – I think I had to less coffee this morning (already updated)

**Review by eduardoboucas — DISMISSED** [review]
LGTM!

**Comment by lukasholzer** [comment]
@eduardoboucas seems that some tests are failing need to investigate further maybe I broke something with the change

File: src/lib/functions/server.test.js [hunk 1]
```diff
@@ -17,6 +18,9 @@ const functionsPath = `functions`
 /** @type { express.Express} */
 let app
 
+// mock the file-system to not write on the real one with `mkdir` inside the registry.js
+mock({ node_modules: mock.load(join(__dirname, '../../../node_modules')) })
```
Comment (lukasholzer) [inline src/lib/functions/server.test.js:22]: tests were failing before as we were doing an mkdir un-mocked on the real fs which is not possible to write inside `/my-functions/functions` on CI therefore I mocked the entire filesystem as the test does not need one.

File: src/lib/functions/registry.js [hunk 2]
```diff
@@ -143,7 +144,10 @@ class FunctionsRegistry {
     log(`${NETLIFYDEVLOG} ${chalk.green('Loaded')} function ${chalk.yellow(name)}.`)
   }
 
-  async scan(directories) {
+  async scan(relativeDirs) {
+    const directories = relativeDirs.filter(Boolean).map((dir) => (isAbsolute(dir) ? dir : join(this.projectRoot, dir)))
```
Comment (lukasholzer) [inline src/lib/functions/registry.js:148]: @erezrokah @ehmicky I think this is the fastet way to fix it -> Wondering if we can fix it on the long term inside the netlify config.

Comment (erezrokah) [reply src/lib/functions/registry.js:148]: Yes, this should be address by https://github.com/netlify/build/issues/3620.

**Review by eduardoboucas — APPROVED** [review]

**Review by erezrokah — APPROVED** [review]

**Pull request merged**